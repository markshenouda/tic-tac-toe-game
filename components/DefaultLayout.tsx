import React from "react";
import { SafeAreaView, StyleSheet } from "react-native";

import { ThemedView } from "@/components/ThemedView";

interface DefaultLayoutProps {
  children: React.ReactNode;
}

export default function DefaultLayout({ children }: DefaultLayoutProps) {
  return (
    <ThemedView style={styles.container}>
      <SafeAreaView style={styles.safeAreaViewStyle}>{children}</SafeAreaView>
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  container: {
    height: "100%",
    padding: 10,
  },
  safeAreaViewStyle: {
    height: "100%",
  },
});
