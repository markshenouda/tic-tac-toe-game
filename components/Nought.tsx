import Svg, { Circle } from "react-native-svg";

export default function Nought() {
  return (
    <Svg width="100%" height="100%" viewBox="0 0 24 24" fill="none">
      <Circle
        cx="12"
        cy="12"
        r="9"
        stroke="white"
        strokeWidth="2.5"
        fill="none"
      />
    </Svg>
  );
}
