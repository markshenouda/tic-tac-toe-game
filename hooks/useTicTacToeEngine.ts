import { useEffect, useRef, useState } from "react";

export enum Player {
  Human = "X",
  Computer = "O",
  None = "",
}

export type Board = Player[][];

export type GameState = {
  board: Board;
  isLoading: boolean;
  winner: Player | "Draw" | null;
};

const emptyBoard = (): Board =>
  Array(3)
    .fill(null)
    .map(() => Array(3).fill(Player.None));

const checkWinner = (board: Board): Player | "Draw" | null => {
  const lines = [
    // Rows
    [
      [0, 0],
      [0, 1],
      [0, 2],
    ],
    [
      [1, 0],
      [1, 1],
      [1, 2],
    ],
    [
      [2, 0],
      [2, 1],
      [2, 2],
    ],
    // Columns
    [
      [0, 0],
      [1, 0],
      [2, 0],
    ],
    [
      [0, 1],
      [1, 1],
      [2, 1],
    ],
    [
      [0, 2],
      [1, 2],
      [2, 2],
    ],
    // Diagonals
    [
      [0, 0],
      [1, 1],
      [2, 2],
    ],
    [
      [0, 2],
      [1, 1],
      [2, 0],
    ],
  ];

  for (const line of lines) {
    const [a, b, c] = line;
    if (
      board[a[0]][a[1]] !== Player.None &&
      board[a[0]][a[1]] === board[b[0]][b[1]] &&
      board[a[0]][a[1]] === board[c[0]][c[1]]
    ) {
      return board[a[0]][a[1]];
    }
  }

  return board.flat().includes(Player.None) ? null : "Draw";
};

const minimax = (
  board: Board,
  depth: number,
  isMaximizing: boolean,
): number => {
  const winner = checkWinner(board);
  if (winner === Player.Computer) return 10 - depth;
  if (winner === Player.Human) return depth - 10;
  if (winner === "Draw") return 0;

  const scores: number[] = [];
  for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
      if (board[i][j] === Player.None) {
        board[i][j] = isMaximizing ? Player.Computer : Player.Human;
        scores.push(minimax(board, depth + 1, !isMaximizing));
        board[i][j] = Player.None;
      }
    }
  }
  return isMaximizing ? Math.max(...scores) : Math.min(...scores);
};

const bestMove = (board: Board): [number, number] => {
  let bestScore = -Infinity;
  let move: [number, number] = [-1, -1];

  for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
      if (board[i][j] === Player.None) {
        board[i][j] = Player.Computer;
        const score = minimax(board, 0, false);
        board[i][j] = Player.None;
        if (score > bestScore) {
          bestScore = score;
          move = [i, j];
        }
      }
    }
  }
  return move;
};

export const useTicTacToe = (playerFirst: boolean) => {
  const [state, setState] = useState<GameState>({
    board: emptyBoard(),
    isLoading: !playerFirst,
    winner: null,
  });

  const timerRef = useRef<NodeJS.Timeout>();

  useEffect(() => {
    if (!playerFirst) {
      timerRef.current = setTimeout(() => makeComputerMove(), 1000);
    }

    return () => clearTimeout(timerRef.current);
  }, [timerRef, playerFirst]);

  const makeMove = (row: number, col: number) => {
    if (
      state.board[row][col] !== Player.None ||
      state.isLoading ||
      state.winner
    )
      return;

    const newBoard = state.board.map((row) => [...row]);
    newBoard[row][col] = Player.Human;
    const winner = checkWinner(newBoard);

    setState({ board: newBoard, isLoading: !winner, winner });

    if (!winner) {
      setTimeout(() => makeComputerMove(), 1000);
    }
  };

  const makeComputerMove = () => {
    setState((prev) => {
      if (prev.winner) return prev;

      const move = bestMove(prev.board);
      if (move[0] === -1) return prev;

      const newBoard = prev.board.map((row) => [...row]);
      newBoard[move[0]][move[1]] = Player.Computer;
      const winner = checkWinner(newBoard);

      return { board: newBoard, isLoading: false, winner };
    });
  };

  return { state, makeMove };
};
