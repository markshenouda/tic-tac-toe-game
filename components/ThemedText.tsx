import { StyleSheet, Text, type TextProps } from "react-native";

import { useThemeColor } from "@/hooks/useThemeColor";

export type ThemedTextProps = TextProps & {
  lightColor?: string;
  darkColor?: string;
  type?: "default" | "title" | "defaultSemiBold" | "subtitle";
};

export function ThemedText({
  style,
  lightColor,
  darkColor,
  type = "default",
  ...rest
}: ThemedTextProps) {
  const color = useThemeColor({ light: lightColor, dark: darkColor }, "text");

  return (
    <Text
      style={[
        { color },
        type === "default" ? styles.default : undefined,
        type === "title" ? styles.title : undefined,
        type === "defaultSemiBold" ? styles.defaultSemiBold : undefined,
        type === "subtitle" ? styles.subtitle : undefined,
        style,
      ]}
      {...rest}
    />
  );
}

const styles = StyleSheet.create({
  default: {
    fontSize: 24,
    lineHeight: 32,
    fontFamily: "DinNextRounded",
  },
  defaultSemiBold: {
    fontSize: 24,
    lineHeight: 32,
    fontWeight: "600",
    fontFamily: "DinNextRounded",
  },
  title: {
    fontSize: 48,
    fontWeight: "bold",
    lineHeight: 58,
    fontFamily: "FeatherBold",
  },
  subtitle: {
    fontSize: 32,
    fontWeight: "bold",
    fontFamily: "DinNextRounded",
  },
});
