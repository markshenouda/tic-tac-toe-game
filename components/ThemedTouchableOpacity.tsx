import { Audio } from "expo-av";
import * as Haptics from "expo-haptics";
import { useEffect, useState } from "react";
import { TouchableOpacity, type TouchableOpacityProps } from "react-native";

import { useThemeColor } from "@/hooks/useThemeColor";

export type ThemedTouchableOpacityProps = TouchableOpacityProps & {
  lightColor?: string;
  darkColor?: string;
};

export function ThemedTouchableOpacity({
  style,
  lightColor,
  darkColor,
  onPressIn,
  onPress,
  ...otherProps
}: ThemedTouchableOpacityProps) {
  const backgroundColor = useThemeColor(
    { light: lightColor, dark: darkColor },
    "background",
  );
  const [sound, setSound] = useState<Audio.Sound | null>(null);

  async function playSound() {
    const { sound } = await Audio.Sound.createAsync(
      require("@/assets/audios/tap.wav"),
    );
    setSound(sound);

    await sound.playAsync();
  }

  useEffect(() => {
    return sound
      ? () => {
          sound.unloadAsync();
        }
      : undefined;
  }, [sound]);

  return (
    <TouchableOpacity
      style={[{ backgroundColor }, style]}
      {...otherProps}
      onPressIn={(event) => {
        if (process.env.EXPO_OS === "ios") {
          // Add a soft haptic feedback when pressing down on the tabs.
          Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
        }
        onPressIn?.(event);
      }}
      onPress={(event) => {
        playSound();
        onPress?.(event);
      }}
    />
  );
}
