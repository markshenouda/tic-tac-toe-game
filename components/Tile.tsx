import { StyleSheet } from "react-native";
import Animated, {
  Easing,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from "react-native-reanimated";

import Cross from "@/components/Cross";
import Nought from "@/components/Nought";
import { ThemedTouchableOpacity } from "@/components/ThemedTouchableOpacity";
import { useThemeColors } from "@/hooks/useThemeColors";
import { Player } from "@/hooks/useTicTacToeEngine";

interface TileProps {
  tileValue: Player;
  onPress: () => void;
  disabled?: boolean;
  accessibilityLabel: string;
  isWinningTile?: boolean;
}

export default function Tile({
  tileValue,
  onPress,
  disabled,
  accessibilityLabel,
  isWinningTile,
}: TileProps) {
  const colors = useThemeColors();
  const crossColor = colors.blue;
  const noughtColor = colors.purple;

  const tileColor = () => {
    if (isWinningTile) {
      return colors.green;
    }

    switch (tileValue) {
      case "X":
        return crossColor;
      case "O":
        return noughtColor;
      default:
        return colors.orange;
    }
  };

  const handlePress = () => {
    if (!disabled) {
      onPress();
    }
  };

  return (
    <ThemedTouchableOpacity
      style={styles.tile}
      lightColor={tileColor()}
      darkColor={tileColor()}
      activeOpacity={0.8}
      focusable={true}
      onPress={handlePress}
      accessible={true}
      accessibilityLabel={accessibilityLabel}
    >
      {tileValue === "X" ? <Cross /> : null}
      {tileValue === "O" ? <Nought /> : null}
    </ThemedTouchableOpacity>
  );
}

const styles = StyleSheet.create({
  tile: {
    aspectRatio: 1,
    margin: 2,
    width: "100%",
    height: "31%",
    borderRadius: 14,
    boxShadow: "0 2 2 rgba(0, 0, 0, 0.25)",
  },
});
