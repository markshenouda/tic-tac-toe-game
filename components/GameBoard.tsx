import React from "react";
import { Platform, StyleSheet } from "react-native";

import { ThemedView } from "@/components/ThemedView";
import Tile from "@/components/Tile";
import { useThemeColors } from "@/hooks/useThemeColors";
import { GameState, Player } from "@/hooks/useTicTacToeEngine";

const BOARD_SIZE = 3;

interface GameBoardProps {
  state: GameState;
  makeMove: (row: number, col: number) => void;
}

export default function GameBoard({ state, makeMove }: GameBoardProps) {
  const colors = useThemeColors();
  return (
    <ThemedView
      style={styles.container}
      lightColor={colors.orangeTint}
      darkColor={colors.orangeTint}
    >
      <ThemedView style={styles.grid}>
        {[...Array(BOARD_SIZE)].map((_, i) => (
          <ThemedView key={i} style={styles.row}>
            {[...Array(BOARD_SIZE)].map((_, j) => (
              <Tile
                key={j}
                tileValue={state.board[i][j]}
                isWinningTile={
                  state.winningLine
                    ? state.winningLine.some(
                        (winningTile) =>
                          winningTile[0] === i && winningTile[1] === j,
                      )
                    : undefined
                }
                onPress={() => {
                  makeMove(i, j);
                }}
                accessibilityLabel={`${state.board[i][j] === Player.Computer ? "Computer" : state.board[i][j] === Player.Human ? "You" : "Empty"}, row ${i + 1}, column ${j + 1}`}
                disabled={
                  state.board[i][j] !== "" ||
                  state.isLoading ||
                  Boolean(state.winner)
                }
              />
            ))}
          </ThemedView>
        ))}
      </ThemedView>
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  container: {
    justifyContent: "center",
    alignItems: "center",
    padding: 8,
    margin: 8,
    marginLeft: "auto",
    marginRight: "auto",
    borderRadius: 12,
    aspectRatio: 1,
    width: "90%",
    maxWidth: 600,
    maxHeight: Platform.OS === "android" && Platform.isTV ? 300 : 600,
  },
  tile: {
    aspectRatio: 1,
    margin: 2,
    width: "100%",
    height: "31%",
    borderRadius: 10,
  },
  grid: {
    flexDirection: "row",
    width: "100%",
    height: "100%",
    backgroundColor: "transparent",
    justifyContent: "space-between",
    gap: "2%",
  },
  row: {
    flexDirection: "column",
    justifyContent: "space-between",
    backgroundColor: "transparent",
    gap: "2%",
  },
});
