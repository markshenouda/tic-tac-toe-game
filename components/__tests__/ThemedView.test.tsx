import { render } from "@testing-library/react-native";
import * as React from "react";

import { ThemedView } from "@/components/ThemedView";
import { useThemeColor } from "@/hooks/useThemeColor";

// Mock the useThemeColor hook
jest.mock("@/hooks/useThemeColor");

describe("ThemedView", () => {
  const mockUseThemeColor = useThemeColor as jest.Mock;

  beforeEach(() => {
    mockUseThemeColor.mockReturnValue("black");
  });

  it("renders correctly with default background color", () => {
    const { toJSON } = render(<ThemedView />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("applies custom light and dark colors", () => {
    mockUseThemeColor.mockReturnValue("customColor");

    const { getByTestId } = render(
      <ThemedView
        lightColor="lightColor"
        darkColor="darkColor"
        testID="themed-view"
      />,
    );
    const viewElement = getByTestId("themed-view");

    expect(viewElement.props.style).toContainEqual({
      backgroundColor: "customColor",
    });
  });
});
