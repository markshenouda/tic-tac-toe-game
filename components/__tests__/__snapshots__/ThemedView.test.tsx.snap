// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedView renders correctly with default background color 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "black",
      },
      undefined,
    ]
  }
/>
`;
